internal/worker/apiaddresssetter/worker.go

```diff
@@ -0,0 +1,366 @@
+			ShouldRestart: func(error) bool { return false },
+			Logger:        internalworker.WrapLogger(config.Logger),
+		}),
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{w.runner},
+	})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	return w, nil
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *apiAddressSetterWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *apiAddressSetterWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *apiAddressSetterWorker) loop() error {
+	controllerNodeChanges, err := w.watchForControllerNodeChanges()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	ctx, cancel := w.scopedContext()
+	defer cancel()
+
+	configChanges, err := w.watchForConfigChanges(ctx)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	for {
+		w.config.Logger.Tracef(ctx, "waiting for controller nodes or addresses changes")
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case <-controllerNodeChanges:
+			// A controller was added or removed.
+			w.config.Logger.Tracef(ctx, "<-controllerChanges")
+			changed, err := w.updateControllerNodes(ctx)
+			if err != nil {
+				return errors.Capture(err)
+			}
+			if !changed {
+				continue
+			}
+			w.config.Logger.Tracef(ctx, "controller node added or removed")
+		case <-w.controllerNodeChanges:
+			// One of the controller nodes changed.
+			w.config.Logger.Tracef(ctx, "<-w.controllerChanges")
+		case <-configChanges:
+			// Controller config has changed.
+			w.config.Logger.Tracef(ctx, "<-w.configChanges")
+
+			// If a config change wakes up the loop before the topology has
+			// been represented in the worker's controller trackers, ignore it;
+			// errors will occur when trying to determine peer group changes.
+			// Continuing is OK because subsequent invocations of the loop will
+			// pick up the most recent config from state anyway.
+			if len(w.controllerTrackers) == 0 {
+				w.config.Logger.Tracef(ctx, "no controller information, ignoring config change")
+				continue
+			}
+		}
+
+		if err := w.updateAPIAddresses(ctx); err != nil {
+			w.config.Logger.Errorf(ctx, "cannot update api addresses: %v", err)
+		}
+	}
+}
+
+func (w *apiAddressSetterWorker) scopedContext() (context.Context, context.CancelFunc) {
+	return context.WithCancel(w.catacomb.Context(context.Background()))
+}
+
+// watchForControllerChanges starts a watcher for changes to controller nodes.
+// It returns a channel which will receive events if any of the watchers fires.
+func (w *apiAddressSetterWorker) watchForControllerNodeChanges() (<-chan struct{}, error) {
+	watcher, err := w.config.ControllerNodeService.WatchControllerNodes()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	if err := w.catacomb.Add(watcher); err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	return watcher.Changes(), nil
+}
```

Comment:
No need to consume the initial event here?

Reply:
What I did is that the first event here will trigger the tracker(s) for the existent controller nodes, so the initial event has relevance.